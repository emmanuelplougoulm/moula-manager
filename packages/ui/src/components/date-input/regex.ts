export const dayFormat = new RegExp(/^(\d\d)$/g);
export const dayMonthFormat = new RegExp(/^(\d\d)\/(\d\d)$/g);
