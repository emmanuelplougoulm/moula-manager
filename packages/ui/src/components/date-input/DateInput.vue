<template>
  <div class="container">
    <LabelItem
      :text="label"
      :highlighted="isHighlighted"
      :tooltip-title="tooltipTitle"
      :tooltip-description="tooltipDescription"
    />
    <input
      :value="formattedDate"
      :placeholder="placeholder"
      data-testid="date-input"
      :class="inputClass"
      class="input input--default text-big-regular"
      maxlength="10"
      @input="onChange"
      @blur="validateInput"
      @focus="resetValidation"
    />
    <ErrorItem :text="errorText" />
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import ErrorItem from '../error/ErrorItem.vue';
import LabelItem from '../label/LabelItem.vue';
import { getErrorText } from '../../utils/commons';
import { dayFormat, dayMonthFormat } from './regex';
import dayjs from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat';

dayjs.extend(customParseFormat);

const props = defineProps({
  label: {
    type: String,
    required: true
  },
  placeholder: {
    type: String,
    default: ''
  },
  modelValue: {
    type: Object,
    required: true
  },
  tooltipTitle: {
    type: String,
    default: undefined
  },
  tooltipDescription: {
    type: String,
    default: undefined
  },
  required: {
    type: Boolean,
    default: false
  }
});

const formattedDate = ref(props.modelValue.value);
const errorText = ref('');
const isValid = ref(true);

const isHighlighted = computed(() => {
  const isMissingAndRequired = !formattedDate.value && props.required;
  return !isValid.value || isMissingAndRequired;
});
const inputClass = computed(() => ({
  ['input--highlighted']: isHighlighted.value
}));

const updateErrorText = () => {
  errorText.value = isHighlighted.value ? getErrorText(props.required, formattedDate.value) : '';
};
const resetValidation = () => {
  isValid.value = true;
};
const validateInput = () => {
  if (formattedDate.value) {
    isValid.value = dayjs(formattedDate.value, 'DD/MM/YYYY', true).isValid();
  }
  updateErrorText();
};
const formatDate = (value: string) => {
  if (!value) {
    resetValidation();
    return '';
  } else if (dayFormat.test(value) || dayMonthFormat.test(value)) {
    return value + '/';
  } else {
    return value;
  }
};
const emit = defineEmits(['update:modelValue']);
const onChange = (event: any): void => {
  const value = event.target.value;
  formattedDate.value = formatDate(value);
  if (value.length === 10) {
    validateInput();
    emit('update:modelValue', { value, isValid: isValid.value });
  }
  updateErrorText();
};
</script>

<style scoped>
.container {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

.input {
  width: 10rem;
}

input:focus-visible {
  outline: 0;
  background-color: var(--neutral-70);
  border: none;
  overflow: hidden;
  box-shadow: none;
}

.input--default {
  height: 3rem;
  padding: 0.75rem 1rem;
  border-radius: 0.25rem;
  border: none;
  background-color: var(--neutral-80);
  box-sizing: border-box;
}

.input--highlighted {
  border: 0.125rem var(--primary-30) solid;
  background-color: var(--primary-90);
  color: var(--primary-30);
}
</style>
