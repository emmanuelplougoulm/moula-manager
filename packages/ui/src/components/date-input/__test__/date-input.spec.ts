import { mount } from '@vue/test-utils';
import DateInput from '../DateInput.vue';
import { dayFormat, dayMonthFormat } from './../regex';

const dateInput = "[data-testid='date-input']";
const inputEvent = 'update:modelValue';

describe('DateInput tests', () => {
  it('should display the default class by default', () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '12/07/1998', isValid: undefined }
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the default class if the value is valid even if the required prop is true', () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '12/07/1998', isValid: undefined },
        required: true
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the highlighted class if there is no value and the required prop is true', () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined },
        required: true
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the highlighted class if the value is invalid', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined }
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('31/02/1998');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a formatted day date if valid', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined }
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('12');
    const formattedDate = input.element.value.replace(dayFormat, '$1/');
    expect(formattedDate).toBe('12/');
  });

  it('should display a formatted day-month date if valid', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined }
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('12/01');
    const formattedDate = input.element.value.replace(dayMonthFormat, '$1/$2/');
    expect(formattedDate).toBe('12/01/');
  });

  it('should emit an event if the full date is in a valid format', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined }
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('12/01/1998');
    expect(wrapper.emitted()).toHaveProperty(inputEvent);
  });

  it('should display a format error if the full date is in an invalid format', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined }
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('31/02/////');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a format error if the full date is a non-existing date', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined }
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('31/02/1998');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a required error if the input is completed then emptied and the required prop is true', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined },
        required: true
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('1');
    await input.setValue('');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a format error if the input is not fully completed and unfocused', async () => {
    const wrapper = mount(DateInput, {
      props: {
        label: 'Label',
        modelValue: { value: '', isValid: undefined }
      }
    });
    const input = wrapper.find<HTMLInputElement>(dateInput);
    await input.setValue('1');
    await input.trigger('blur');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
