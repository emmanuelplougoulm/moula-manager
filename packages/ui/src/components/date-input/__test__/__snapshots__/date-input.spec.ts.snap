// Vitest Snapshot v1

exports[`DateInput tests > should display a format error if the full date is a non-existing date 1`] = `"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input--highlighted input input--default text-big-regular\\" maxlength=\\"10\\"><span data-v-c1d8433f=\\"\\" data-v-c69729d6=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Invalid format</span></div>"`;

exports[`DateInput tests > should display a format error if the full date is in an invalid format 1`] = `"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input--highlighted input input--default text-big-regular\\" maxlength=\\"10\\"><span data-v-c1d8433f=\\"\\" data-v-c69729d6=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Invalid format</span></div>"`;

exports[`DateInput tests > should display a format error if the input is not fully completed and unfocused 1`] = `"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input--highlighted input input--default text-big-regular\\" maxlength=\\"10\\"><span data-v-c1d8433f=\\"\\" data-v-c69729d6=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Invalid format</span></div>"`;

exports[`DateInput tests > should display a required error if the input is completed then emptied and the required prop is true 1`] = `"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input--highlighted input input--default text-big-regular\\" maxlength=\\"10\\"><span data-v-c1d8433f=\\"\\" data-v-c69729d6=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Required</span></div>"`;

exports[`DateInput tests > should display the default class by default 1`] = `
"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input input--default text-big-regular\\" maxlength=\\"10\\">
  <!--v-if-->
</div>"
`;

exports[`DateInput tests > should display the default class if the value is valid even if the required prop is true 1`] = `
"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input input--default text-big-regular\\" maxlength=\\"10\\">
  <!--v-if-->
</div>"
`;

exports[`DateInput tests > should display the highlighted class if the value is invalid 1`] = `"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input--highlighted input input--default text-big-regular\\" maxlength=\\"10\\"><span data-v-c1d8433f=\\"\\" data-v-c69729d6=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Invalid format</span></div>"`;

exports[`DateInput tests > should display the highlighted class if there is no value and the required prop is true 1`] = `
"<div data-v-c69729d6=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-c69729d6=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span><input data-v-c69729d6=\\"\\" placeholder=\\"\\" data-testid=\\"date-input\\" class=\\"input--highlighted input input--default text-big-regular\\" maxlength=\\"10\\">
  <!--v-if-->
</div>"
`;
